### PR TITLE
Initial work on Traversal Improvements in GraphqlSchema builds

### PR DESCRIPTION
**Pull request by bbakerman**
This is a WORK implementation of building a GraQhlSchema object with only less traversal

At this stage I left the OLD code in there so you could see what's different

It would be cleaned up if we accepted this code.

I copied the bench mark from the previous PR and run it.

It shows that this code as is BETTER - but not ground breakingly better.

Start

```
Benchmark                                    Mode  Cnt   Score   Error  Units
SchemaBenchMark.benchMarkLargeSchemaCreate  thrpt   15  18.134 ± 1.099  ops/s
```

This work as is

```
Benchmark                                    Mode  Cnt   Score   Error  Units
SchemaBenchMark.benchMarkLargeSchemaCreate  thrpt   15  19.771 ± 0.685  ops/s
```

I then took out the validation steps just to see what would happen to the numbers.

```
Benchmark                                    Mode  Cnt   Score   Error  Units
SchemaBenchMark.benchMarkLargeSchemaCreate  thrpt   15  25.906 ± 0.370  ops/s
```

I then took out the type replacement step and validation just to see.  This is not valid of course

```
Benchmark                                    Mode  Cnt   Score   Error  Units
SchemaBenchMark.benchMarkLargeSchemaCreate  thrpt   15  30.256 ± 1.171  ops/s
```

Having the ability to skip validation would help.  Type replacements MUST happen.

If we could find a way to combine type replacement with the previous steps then it would be better.  But that might 
be impossible - we need ALL the types collected to do type ref replacements... OR do we. ???   Anyway that's for another PR based on this one

### CODE REVIEW

File: src/main/java/graphql/schema/CodeRegistryVisitor.java [hunk 1]
```diff
@@ -12,11 +23,12 @@
  * This ensure that all fields have data fetchers and that unions and interfaces have type resolvers
```
Comment (bbakerman) [inline src/main/java/graphql/schema/CodeRegistryVisitor.java:23]: I moved this to a schem.impl package just to clean things up.  Its kinda of a private thing and not general purpose

File: src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java [hunk 2]
```diff
@@ -1,12 +1,26 @@
-package graphql.schema;
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java:1]: Again I moved this to a schem.impl package just to clean things up. Its kinda of a private thing and not general purpose

File: src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+import graphql.schema.GraphQLTypeReference;
+import graphql.schema.GraphQLTypeVisitor;
+import graphql.schema.GraphQLUnionType;
+import graphql.schema.GraphQLUnmodifiedType;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+
+import java.util.List;
+
+public class MultiReadOnlyGraphQLTypeVisitor implements GraphQLTypeVisitor {
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java:34]: Allows us to call N visitors at once - read only ones that can change nodes

File: src/main/java/graphql/schema/impl/SchemaUtil.java [hunk 4]
```diff
@@ -1,11 +1,24 @@
-package graphql.schema;
+package graphql.schema.impl;
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/SchemaUtil.java:1]: Again again I moved this to a schem.impl package just to clean things up. Its kinda of a private thing and not general purpose

File: src/main/java/graphql/schema/impl/StronglyConnectedComponentsTopologicallySorted.java [hunk 5]
```diff
@@ -1,7 +1,9 @@
-package graphql.schema;
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/StronglyConnectedComponentsTopologicallySorted.java:1]: Again again again I moved this to a schem.impl package just to clean things up. Its kinda of a private thing and not general purpose

**Review by andimarek — APPROVED** [review]

File: src/main/java/graphql/schema/GraphQLSchema.java [hunk 6]
```diff
@@ -690,12 +759,42 @@ GraphQLSchema buildImpl(boolean afterTransform) {
...
+            ImmutableMap<String, GraphQLNamedType> allTypes = typeCollectingVisitor.getResult();
+            List<GraphQLNamedType> allTypesAsList = getAllTypesAsList(allTypes);
+            codeRegistry = extractedDataFetchers.build();
+
+            ImmutableMap<String, List<GraphQLObjectType>> groupedImplementations = schemaUtil.groupInterfaceImplementationsByName(allTypesAsList);
+            ImmutableMap<String, ImmutableList<GraphQLObjectType>> interfaceNameToObjectTypes = buildInterfacesToObjectTypes(groupedImplementations);
+
+            final GraphQLSchema finalSchema = new GraphQLSchema(partiallyBuiltSchema, codeRegistry, allTypes, interfaceNameToObjectTypes);
+            schemaUtil.replaceTypeReferences(finalSchema);
+            if (true) {
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLSchema.java:779]: this doesn't look right

File: src/main/java/graphql/schema/GraphQLSchema.java [hunk 7]
```diff
@@ -68,6 +70,67 @@
...
+        this.interfaceNameToObjectTypeNames = ImmutableKit.emptyMap();
+    }
+
+    /*
+     * This constructs a full fledged graphql schema object that has not yet had its type references replaced
+     * but its otherwise complete
+     */
+    @Internal
+    public GraphQLSchema(GraphQLSchema partiallyBuiltSchema,
+                         GraphQLCodeRegistry codeRegistry, ImmutableMap<String, GraphQLNamedType> typeMap,
```
Comment (andimarek) [inline src/main/java/graphql/schema/GraphQLSchema.java:113]: it should be one line per arg, right?

File: src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java (hunk 3 above)
Comment (andimarek) [reply src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java:34]: Is that public api?

Comment (andimarek) [reply src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java:34]: I am not so sure about the semantics here: they don't compose completely, right? I can't return a ABORT/Quit anymore, for example.

**Review by andimarek — CHANGES_REQUESTED** [review]

File: src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java (hunk 3 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/impl/MultiReadOnlyGraphQLTypeVisitor.java:40]: They dont compose generically  - they run one after the other.  I have made it `@Internal` and improved the javadoc on what it does

File: src/main/java/graphql/schema/GraphQLSchema.java (hunk 6 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/GraphQLSchema.java]: Yeah as discussed direct - this was to show the new code versus the old.

The PR now has this removed and cleaned up

File: src/main/java/graphql/schema/GraphQLSchema.java (hunk 7 above)
Comment (bbakerman) [reply src/main/java/graphql/schema/GraphQLSchema.java]: done

File: src/main/java/graphql/schema/impl/SchemaUtil.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                    List<GraphQLImplementingType> myGroup = result.computeIfAbsent(interfaceType.getName(), k -> new ArrayList<>());
+                    myGroup.add((GraphQLImplementingType) type);
+                }
+            }
+        }
+        return ImmutableMap.copyOf(new TreeMap<>(result));
+    }
+
+    public static void replaceTypeReferences(GraphQLSchema schema) {
+        final Map<String, GraphQLNamedType> typeMap = schema.getTypeMap();
```
Comment (andimarek) [inline src/main/java/graphql/schema/impl/SchemaUtil.java:93]: I am not sure about this code: it looks suspicious that we grab all types to iterate. I think we can do better and maybe this is not even completely correct. But we can leave it for another PR.

**Review by andimarek — APPROVED** [review]
